docs/ovs-pipeline.md

```diff
@@ -0,0 +1,561 @@
+3. table=50, priority=200,ip,nw_src=10.10.1.3 actions=conjunction(2,1/3)
+4. table=50, priority=200,ip,nw_dst=10.10.1.2 actions=conjunction(2,2/3)
+5. table=50, priority=200,ip,nw_dst=10.10.1.3 actions=conjunction(2,2/3)
+6. table=50, priority=200,tcp,tp_dst=80 actions=conjunction(2,3/3)
+7. table=50, priority=190,conj_id=2,ip actions=resubmit(,70)
+8. table=50, priority=80,ip actions=resubmit(,60)
+```
+
+Notice how we use the OVS built-in `conjunction` action to implement policies
+efficiently. This enables us to do a conjunctive match across multiple
+dimensions (source IP, destination IP, port) efficiently without "exploding" the
+number of flows. By definition of a conjunctive match, we have at least 2
+dimensions. For our use-case we have at most 3 dimensions.
+
+The above example flows read as follow: if the source IP address is in set
+{10.10.1.2, 10.10.1.3}, and the destination IP address is in the set {10.10.1.2,
+10.10.1.3}, and the destination TCP port is in the set {80}, then use the
+`conjunction` action with id 2, which is resubmit to
+[L3ForwardingTable]. Otherwise, resubmit to [EgressDefaultTable].
+
+The only requirements on `conj_id` is for it to be a unique 32-bit integer
+within the table. At the moment we use a single custom allocator, which is
+common to [EgressRuleTable] and [IngressRuleTable]. This is why `conj_id` is set
+to 2 in the above example (1 was allocated for the ingress rule of our Network
+Policy example).
+
+If the Network Policy specification includes exceptions (`except` field), then
+the table will include additional rules, but we will not cover them in this
+document.
+
+If the `conjunction` action is matched, packets are "allowed" and resubmitted
+directly to [L3ForwardingTable]. Other packets go to [EgressDefaultTable]. If a
+connection is established - as a remainder all connections are committed in
+[ConntrackStateTable] - its packets go straight to [L3ForwardingTable], with no
+other match required (see flow 1 above, which has the highest priority). This
+means that ongoing connections are not affected if the K8s Network Policies are
+updated.
+
+One thing to keep in mind is that for service traffic, these rules are applied
+after the packets have gone through the local gateway and through kube-proxy. At
+this point the ingress port is no longer the Pod port, but the local gateway
+port. Therefore we cannot use the port as the match condition to identify if the
+Pod has been applied a Network Policy - which is what we do for the
+[IngressRuleTable] -, but instead have to use the source IP address.
+
+### EgressDefaultTable (60)
+
+This table complements [EgressRuleTable] for Network Policy egress rule
+implementation. In K8s, when a Network Policy is applied to a set of Pods, the
+default behavior for these Pods become "deny" (it becomes an [isolated
+Pod](https://kubernetes.io/docs/concepts/services-networking/network-policies/#isolated-and-non-isolated-pods). This
```

Comment:
```suggestion
Pod](https://kubernetes.io/docs/concepts/services-networking/network-policies/#isolated-and-non-isolated-pods)). This
```
Otherwise it appears as `... "deny" (it becomes an isolated Pod. This ...`